### PR TITLE
issue #66 Add Kata Diamond

### LINKED ISSUE #66: [TDD] Kata - Diamond
**Description**
The requirement for this issue is to print a diamond
[Kata-Dimond](https://codingdojo.org/kata/Diamond/)

you can follow this reference to enjoy TDD. Fail-> Pass -> Fail -> Pass
[Reference](http://claysnow.co.uk/recycling-tests-in-tdd/) 

We would like to make it simpler to check its repeat characters.
```
Given 1 parameters (cahr)
When char = 'A'
Expect output
 A

When char = 'B'
Expect output
 ABBA

When char = 'C'
Expect 
 ABBCCCBBA

When char = 'D'
Expect 
 ABBCCCDDDDCCCBBA

.....  so on , should fit From 'A' to 'Z'

When char = '1'
Expect throw IllegalArgumentExcpetion\



```
**Additional context**

Add any other context, links references or screenshots about the feature request here.

**Important Checklist**
- [ ] I have written unit test first, make it red(fail first).
- [ ] I have then written logic code, make it green(success)
- [ ] I have tried to refactor my code, make it understandable(document my code)

### PR DESCRIPTION
**Pull request by farzado**
* Generate diamond string based on a given wide point character that is in range of A-Z.
* Used TDD approach by adding unit tests first.

Resolves #66

### CODE REVIEW

File: src/test/java/com/ordestiny/tdd/kata/DiamondTest.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	@Test
+	public void buildDiamond_widestPointCharSetToCharB_returnCorrectDiamondString() {
+		assertEquals(Diamond.buildDiamond('B'), "ABBA");
+	}
+
+	@Test
+	public void buildDiamond_widestPointCharSetToCharC_returnCorrectDiamondString() {
+		assertEquals(Diamond.buildDiamond('C'), "ABBCCCBBA");
```
Comment (januslinhc) [inline src/test/java/com/ordestiny/tdd/kata/DiamondTest.java:31]: the logic of buildDiamond_widestPointCharSetToA_returnA, buildDiamond_widestPointCharSetToCharB_returnCorrectDiamondString and buildDiamond_widestPointCharSetToCharC_returnCorrectDiamondString methods are exactly the same.

I think you can use parameterized test instead??

Comment (farzado) [reply src/test/java/com/ordestiny/tdd/kata/DiamondTest.java]: Thanks for your suggestion @januslinhc . I've updated the PR.

**Comment by farzado** [comment]
@fengyuanyang I've updated the PR, could you pls have a look?

**Review by januslinhc — APPROVED** [review]

**Review by fengyuanyang — COMMENTED** [review]
> @fengyuanyang I've updated the PR, could you pls have a look?

I didn't see changes, there is only one commit in this pr for now

**Comment by januslinhc** [comment]
> > @fengyuanyang I've updated the PR, could you pls have a look?
> 
> 
> 
> I didn't see changes, there is only one commit in this pr for now 

Becoz he replaced his original commit 😂

File: src/test/java/com/ordestiny/tdd/kata/DiamondTest.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	public void buildDiamond_widestPointCharSetToCharGreaterThanA_returnedValueHasCorrectNumberOfRepetitions() {
+		for (char widestPointChar = 'B'; widestPointChar <= 'Z'; widestPointChar++) {
+			// In each generated diamond string, the widest point character is repeated one time more than its distance to
+			// the base character
+			int diagonalLength = widestPointChar - Diamond.BASE_CHARACTER + 1;
+			String diamondString = Diamond.buildDiamond(widestPointChar);
+			char[] diagonal = new char[diagonalLength];
+			Arrays.fill(diagonal, widestPointChar);
+			assertTrue(diamondString.contains(String.valueOf(diagonal)));
+			assertEquals(Math.pow(diagonalLength, 2), diamondString.length());
```
[the remark this answers is not in the document]
Comment (farzado) [reply src/test/java/com/ordestiny/tdd/kata/DiamondTest.java]: Thanks for the recommendation, updated the PR as you suggested.

File: src/test/java/com/ordestiny/tdd/kata/DiamondTest.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+	@Test
+	public void buildDiamond_validWidestPointChars_returnCorrectDiamondString() {
+    Map<Character, String> diamonds = new HashMap<>();
+    diamonds.put('A', "A");
+    diamonds.put('B', "ABBA");
+    diamonds.put('C', "ABBCCCBBA");
+    diamonds.put('D', "ABBCCCDDDDCCCBBA");
+		diamonds.put('E', "ABBCCCDDDDEEEEEDDDDCCCBBA");
+    diamonds.forEach(this::verifyDiamondString);
```
[the remark this answers is not in the document]
Comment (farzado) [reply src/test/java/com/ordestiny/tdd/kata/DiamondTest.java]: Noted and corrected in the updated PR.

File: src/main/java/com/ordestiny/tdd/kata/Diamond.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+		StringBuilder leftPart = new StringBuilder();
+		int offset = widestPointChar - BASE_CHARACTER;
+			for(int i=0; i< offset; i++){
+				for(int j=0; j< i+1; j++){
+					leftPart.append ((char) (BASE_CHARACTER + i));
+				}
+			}
+		char[] diagonalString = new char[offset + 1];
+		Arrays.fill(diagonalString, widestPointChar);
+		return leftPart + String.valueOf(diagonalString) + leftPart.reverse();
```
[the remark this answers is not in the document]
Comment (farzado) [reply src/main/java/com/ordestiny/tdd/kata/Diamond.java]: Noted and reflected in the updated PR.

File: src/main/java/com/ordestiny/tdd/kata/Diamond.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+			throw new IllegalArgumentException("Supplied widest point character is not valid. Only A-Z is accepted.");
+		}
+
+		StringBuilder leftPart = new StringBuilder();
+		int offset = widestPointChar - BASE_CHARACTER;
+			for(int i=0; i< offset; i++){
+				for(int j=0; j< i+1; j++){
+					leftPart.append ((char) (BASE_CHARACTER + i));
+				}
+			}
```
[the remark this answers is not in the document]
Comment (farzado) [reply src/main/java/com/ordestiny/tdd/kata/Diamond.java]: Noted and corrected in the updated PR.

**Comment by januslinhc** [comment]
I have no idea why he need force push😂

**Comment by farzado** [comment]
> I have no idea why he need force push😂

Sorry if it has confused use, some people prefer to have a clean linear git history and to squash all commits into one, that's why I used force push.

**Comment by januslinhc** [comment]
> > I have no idea why he need force push😂
> 
> 
> 
> Sorry if it has confused you, some people prefer to have a clean linear git history and to squash all commits into one, that's why I used force push. 

I see. That's why @fengyuanyang asked why some commits lost

**Comment by januslinhc** [comment]
> > > I have no idea why he need force push😂
> > 
> > 
> > 
> > Sorry if it has confused you, some people prefer to have a clean linear git history and to squash all commits into one, that's why I used force push. 
> 
> I see. That's why @fengyuanyang asked why some commits lost

For me, it's fine if the git history is dirty since you are working in your feature sub-branch. If we commit our changes as squash, it's hard for us to revert in case we need to

**Comment by fengyuanyang** [comment]
oh , that's why !! Thanks for your explanation!!
It seems difficult for reviewer to see what has changed with force push . 
It looks good to me now , 
Thanks @januslinhc  and @farzado

**Pull request merged**